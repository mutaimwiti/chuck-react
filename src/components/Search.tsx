import debounce from 'lodash.debounce';
import React, {ChangeEvent, useState, MouseEvent} from "react";
import {Alert, Button, Form, InputGroup, ListGroup} from "react-bootstrap";

import {JokeState, SearchState} from "../store/reducers/joke";

type SearchProps = {
    results: SearchState,
    onSearch(phrase: string): void,
    onClearSearch(): void,
}

const Search: React.FC<SearchProps> = ({results, onSearch, onClearSearch}) => {
    const {items, isLoading} = results;

    const [query, setQuery] = useState('');
    const [valid, setValid] = useState(false);

    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
        const {value} = event.target;
        setQuery(value);

        // search phrase should be a minimum of 3 three characters
        if (value.length >= 3) {
            const debouncedSearch = debounce(() => onSearch(value), 300);
            debouncedSearch();
            setValid(true);
        } else {
            onClearSearch();
            setValid(false);
        }
    }

    const handleClear = (event: MouseEvent<HTMLButtonElement>) => {
        event.currentTarget.blur();
        setQuery('');
        onClearSearch();
    }

    return (
        <>
            <Form>
                <Form.Group>
                    <Form.Label><b>Search more jokes</b></Form.Label>
                    <InputGroup className="mb-3">
                        <Form.Control
                            type="text"
                            value={query}
                            onChange={handleChange}
                            placeholder="Enter your search phrase e.g. food"
                        />
                        <InputGroup.Append>
                            <Button variant="outline-secondary" onClick={handleClear}>Clear</Button>
                        </InputGroup.Append>
                    </InputGroup>
                    {
                        !valid && (<Form.Text muted>
                            Your search phrase must be 3 characters or longer
                        </Form.Text>)
                    }
                </Form.Group>
            </Form>
            <br/>
            {
                isLoading ? <Alert variant="info">Loading jokes...</Alert> : (
                    valid && !items.length && <Alert variant="warning">No results to display</Alert>
                )
            }
            <ListGroup>
                {items.map((joke: JokeState) => {
                    return <ListGroup.Item key={joke.id}>{joke.value}</ListGroup.Item>
                })}
            </ListGroup>
        </>
    );
}

export default Search;
