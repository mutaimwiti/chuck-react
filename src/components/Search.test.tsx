import React from 'react';
import debounce from 'lodash.debounce';
import { fireEvent, render } from '@testing-library/react';

import Search, { SearchProps } from './Search';

jest.mock('lodash.debounce');

describe('Search', function () {
  let props: SearchProps;

  beforeEach(() => {
    // @ts-ignore
    debounce.mockImplementation((fnToRun) => {
      return () => fnToRun();
    });

    props = {
      results: {
        isLoading: false,
        items: [
          {
            icon_url: 'icon/url',
            id: 'someJokeId',
            url: 'joke/url',
            value: 'This is a joke',
          },
          {
            icon_url: 'icon/url',
            id: 'someOtherJokeId',
            url: 'joke/url',
            value: 'This is another joke',
          },
        ],
      },
      onSearch: jest.fn(),
      onClearSearch: jest.fn(),
    };
  });

  it('should not blow up', () => {
    expect(() => render(<Search {...props} />)).not.toThrow();
  });

  it('should match snapshot', () => {
    const { container } = render(<Search {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should display jokes', () => {
    const { getByText } = render(<Search {...props} />);

    expect(getByText('This is a joke')).toBeInTheDocument();
    expect(getByText('This is another joke')).toBeInTheDocument();
  });

  describe('when search query is typed in', () => {
    describe('and it is less than 3 characters', () => {
      it('should show tip and invoke onClearSearch', () => {
        const { getByTestId, getByText } = render(<Search {...props} />);

        expect(props.onClearSearch).not.toHaveBeenCalled();

        const input = getByTestId('search-input');
        fireEvent.change(input, { target: { value: 'fo' } });

        expect(props.onClearSearch).toHaveBeenCalledTimes(1);

        expect(getByText('Your search phrase must be 3 characters or longer'));
      });
    });

    describe('and it more than 2 characters', () => {
      it('should show invoke onSearch with query', () => {
        const { getByTestId } = render(<Search {...props} />);

        const input = getByTestId('search-input');

        expect(props.onSearch).not.toHaveBeenCalled();

        fireEvent.change(input, { target: { value: 'foo' } });

        expect(props.onSearch).toHaveBeenCalledTimes(1);
        expect(props.onSearch).toHaveBeenCalledWith('foo');
      });
    });
  });

  describe('when it is loading jokes', () => {
    it('should show loading message', () => {
      const propsCopy = props;

      propsCopy.results.isLoading = true;

      const { getByText } = render(<Search {...propsCopy} />);

      expect(getByText('Your search phrase must be 3 characters or longer'));
    });
  });

  describe('when there are no results', () => {
    it('should show no results to display message', () => {
      const propsCopy = props;

      propsCopy.results.items = [];
      propsCopy.results.isLoading = false;

      const { getByText, getByTestId } = render(<Search {...propsCopy} />);

      const input = getByTestId('search-input');

      fireEvent.change(input, { target: { value: 'foo' } });

      expect(getByText('No results to display'));
    });
  });

  describe('when clear button is clicked', () => {
    it('should invoke onClearSearch', () => {
      const { getByText } = render(<Search {...props} />);

      expect(props.onClearSearch).not.toHaveBeenCalled();

      const input = getByText('Clear');
      fireEvent.click(input);

      expect(props.onClearSearch).toHaveBeenCalledTimes(1);
    });
  });
});
