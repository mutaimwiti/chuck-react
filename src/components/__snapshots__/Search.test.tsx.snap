// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should match snapshot 1`] = `
<div>
  <form
    class=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
      >
        <b>
          Search more jokes
        </b>
      </label>
      <div
        class="mb-3 input-group"
      >
        <input
          class="form-control"
          data-testid="search-input"
          placeholder="Enter your search phrase e.g. food"
          type="text"
          value=""
        />
        <div
          class="input-group-append"
        >
          <button
            class="btn btn-outline-secondary"
            type="button"
          >
            Clear
          </button>
        </div>
      </div>
      <small
        class="form-text text-muted"
      >
        Your search phrase must be 3 characters or longer
      </small>
    </div>
  </form>
  <br />
  <div
    class="list-group"
  >
    <div
      class="list-group-item"
    >
      This is a joke
    </div>
    <div
      class="list-group-item"
    >
      This is another joke
    </div>
  </div>
</div>
`;
